pkg/workspace/controllers/workspace_controller.go

```diff
@@ -107,6 +107,10 @@ func (c *WorkspaceReconciler) Reconcile(ctx context.Context, req reconcile.Reque
 }
 
 func (c *WorkspaceReconciler) ensureFinalizer(ctx context.Context, workspaceObj *kaitov1beta1.Workspace) error {
+	// Skip adding the finalizer if the workspace is already marked for deletion.
+	if !workspaceObj.DeletionTimestamp.IsZero() {
+		return nil
+	}
 	if !controllerutil.ContainsFinalizer(workspaceObj, consts.WorkspaceFinalizer) {
```

Comment:
The concern here is if the finalizer is removed and then in following reconcile finalizer is re-added.

Reply:
This line seems unnecessary. If the workspace is deleted without finalizer, when you hit line 111, the object is gone in the apiserver and following patch cmd will fail and the next retry will automatically skipped since the object is gone.

Reply:
Could there be a scenario where the object is still present with a deletion timestamp? In this scenario, adding the finalizer again we can avoid. Deletion in Progress vs. Deletion Complete

Reply:
kind of like a defensive programming thing here - wouldn't hurt

Reply:
removed